Allow password encodings in config

tobes opened the pull request:
From #415 this adds universal base64 encoding via
```
module {
   value:base64 = '<encoded>'
}
```

@ultrabug if you are happy with this concept I'll update this PR with some documentation

ultrabug commented on py3status/parse_config.py:
```diff
@@ -441,6 +442,12 @@ def parse(self, dictionary=None, end_token=None):
+                    # indicates that it has been encoded.
+                    if ':' in name:
+                        (name, scheme) = name.split(':')
+                        if scheme == 'base64':
```
if this ought to be extendable, I'd rather have this done in a dedicated method please

ultrabug commented on py3status/parse_config.py:
```diff
@@ -100,7 +101,7 @@ class ConfigParser:
         r'("(?:[^"\\]|\\.)*")'  # double quoted string
         r"|('(?:[^'\\]|\\.)*')"  # single quoted string
-        '|([a-z_][a-z0-9_]*)'  # token
+        '|([a-z_][a-z0-9_]*(:[a-z0-9_]+)?)'  # token
```
I think `a-z0-9_` equivalent is `\w`

tobes replied:
sometimes explicit is good eg is `-` in `\w`

ultrabug replied:
true

ultrabug commented on py3status/parse_config.py:
```diff
@@ -4,6 +4,7 @@
 import glob
 import os
 import imp
+import base64
```
bad kitty

ultrabug commented:
I pretty like the concept of generalization here yes tho I'm still not convinced that base64 encoding anything increases the security of #415

tobes commented:
> I'm still not convinced that base64 encoding anything increases the security

I was off the same opinion but it is still useful.  It means I can look at you config and not just read your password by accident.  But it is completely insecure

We could do something proper secure and have py3status prompt for a pass phrase and not start affected modules till it has one.  But I'm not sure that it is worth the effort at the moment.

tobes commented:
I've moved the processing into it's own method.  It means we can now do some insane stuff in our configs, not that it should be supported.

```
static_string test {
    format:base64 = 'MTIzNA=='
    format += '=='
    format:base64 += 'MTIzNA=='
}
```

ultrabug commented on doc/README.md:
```diff
@@ -7,6 +7,7 @@ py3status documentation
 * [Available modules](#available_modules)
 * [Ordering modules](#ordering_modules)
 * [Configuring modules](#configuring_modules)
+* [Obfuscation](#obfuscation)
```
**Configuration obfuscation** maybe more understandable ?

ultrabug commented on doc/README.md:
```diff
@@ -83,14 +84,34 @@ imap {
+
+__New in version 3.1__
+
+In the previous example configure the users password is in plain text.  Users may
```
... example configuration ...

tobes commented:
Made those doc fixes

tobes commented:
So I was having a think about the whole privacy stuff and decided that it was an interesting topic.

py3status would easily leak any information that you encode eg if you use logging, notifications, put it in your format string.  So I've updated this PR to stop this leakage.  It is not an easy problem as we need to ensure that modules work, especially if they use external libs eg requests.

This implementation allows the value of 'private' config options only to be known to modules with the same name as the config option is given for.  And when used outside this `***` is used as the value.

All unicode methods should be available and work both when valid and not.

The whole thing is academic as all I need is to add a module that monkey patches this out and I can get the real values.  So this is more about keeping things semi-private and a bit of fun.

As well as `:base64` I have added `:hide` that is plain text in the config but will not leak to logs, i3bar etc

tobes commented:
A few more updates.

* We are more resistant to modules trying to change the Private classes

* Only allow this for py3status modules

* Only allow for strings

* Check the 'scheme' exists

ultrabug commented:
I'm tempted to use `:private` instead of `:hide`

Also, since :base64 et al are used to obfuscate, why don't we hide them too by default ?

ultrabug commented:
> I'm tempted to use :private instead of :hide

dunno, :hide seems more explicit after some more thoughts

tobes commented:
> Also, since :base64 et al are used to obfuscate, why don't we hide them too by default ?

We do the :hide is just so I can have my password hidden in the logs all :xxx are hidden in this way.

Do I need to make the docs clearer?

tobes commented:
Hopefully clearer documentation

ultrabug commented:
Thanks this is a neat addition :)
